### PR TITLE
IEP-1018 Automation of hints viewer

### PR DESCRIPTION
**Pull request by sigmaaa**
## Description

Added automated build hints viewer which appears after the build if there are available hints for founded errors. Also, can be turned on manually via Windows -> Show View -> Other... -> Espressif -> Build Hints

Fixes # ([IEP-1018](https://jira.espressif.com:8443/browse/IEP-1018))

## Type of change

Please delete options that are not relevant.
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How has this been tested?

Test 1:
- Add some errors which will match regular expressions from the hints.yml, for example, `#import spiram.h` 
- After build should appear and focused Build Hints viewer with a suggested hint
Test 2:
- turn off search for hints in the Espressif Preference Page -> close Build Hints viewer and rebuild again -> Build Hints should not be opened automatically

**Test Configuration**:
* ESP-IDF Version:
* OS (Windows,Linux and macOS):

## Dependent components impacted by this PR:

- Build 

## Checklist
- [x] PR Self Reviewed
- [x] Applied Code formatting
- [ ] Added Documentation
- [x] Added Unit Test
- [ ] Verified on all platforms - Windows,Linux and macOS

### Summary by CodeRabbit

- New Feature: Added automated build hints functionality to assist users in identifying and resolving ESP-IDF errors. This feature can be enabled or disabled via a new checkbox in the preferences dialog.
- New Feature: Introduced a new UI view, "BuildView", for displaying build hints. The view consists of a table that displays error messages and corresponding hints.
- Refactor: Improved handling of regular expressions in the `ReHintPair` class by using `Pattern` objects instead of `String`.
- Test: Added comprehensive tests for the new `EspIdfErrorParser` class and updated existing tests to accommodate changes in the `ReHintPair` class.
- Chore: Updated various UI elements and tooltips to improve user experience and provide additional information about new features.

### CODE REVIEW

**Review by sigmaaa — COMMENTED** [review]
Self reviewed. Need to add documentation and test coverage for EspIdfErrorParser

**Comment by AndriiFilippov** [comment]
@sigmaaa hi ! 
Tested under: 
OS - Windows 10 / MacOS 
ESP-IDF: v5.0 - v5.1 

release 5.1 / 5.1.1 LGTM 👍  able to enable / disable hint feature. able to see useful hints. 

release 5.0 / 5.0.3 

when hint option is selected 
image

unable to build a project.
the build just stuck. 

 
image




LOGS:
```
-- Check size of time_t - done
-- Found Python3: C:/Python311/python.exe (found version "3.11.0") found components: Interpreter 
-- Performing Test C_COMPILER_SUPPORTS_WFORMAT_SIGNEDNESS
-- Performing Test C_COMPILER_SUPPORTS_WFORMAT_SIGNEDNESS - Success
-- App "hello_world" version: v5.0-dev-4379-g36f49f361-dirty
-- Adding linker script C:/Users/AndriiFilippov/runtime-launchConfiguration(12)/fcvijnl/build/esp-idf/esp_system/ld/memory.ld
-- Adding linker script C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_system/ld/esp32s2/sections.ld.in
-- Adding linker script C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_rom/esp32s2/ld/esp32s2.rom.ld
-- Adding linker script C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_rom/esp32s2/ld/esp32s2.rom.api.ld
-- Adding linker script C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_rom/esp32s2/ld/esp32s2.rom.libgcc.ld
-- Adding linker script C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_rom/esp32s2/ld/esp32s2.rom.newlib-funcs.ld
-- Adding linker script C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_rom/esp32s2/ld/esp32s2.rom.newlib-data.ld
-- Adding linker script C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_rom/esp32s2/ld/esp32s2.rom.spiflash.ld
-- Adding linker script C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/soc/esp32s2/ld/esp32s2.peripherals.ld
-- Components: app_trace app_update bootloader bootloader_support bt cmock console cxx driver efuse esp-tls esp_adc esp_app_format esp_common esp_eth esp_event esp_gdbstub esp_hid esp_http_client esp_http_server esp_https_ota esp_https_server esp_hw_support esp_lcd esp_local_ctrl esp_netif esp_partition esp_phy esp_pm esp_psram esp_ringbuf esp_rom esp_system esp_timer esp_wifi espcoredump esptool_py fatfs freertos hal heap http_parser idf_test ieee802154 json log lwip main mbedtls mqtt newlib nvs_flash openthread partition_table perfmon protobuf-c protocomm pthread sdmmc soc spi_flash spiffs tcp_transport touch_element ulp unity usb vfs wear_levelling wifi_provisioning wpa_supplicant xtensa
-- Component paths: C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/app_trace C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/app_update C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/bootloader C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/bootloader_support C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/bt C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/cmock C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/console C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/cxx C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/driver C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/efuse C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp-tls C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_adc C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_app_format C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_common C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_eth C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_event C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_gdbstub C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_hid C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_http_client C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_http_server C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_https_ota C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_https_server C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_hw_support C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_lcd C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_local_ctrl C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_netif C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_partition C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_phy C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_pm C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_psram C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_ringbuf C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_rom C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_system C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_timer C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esp_wifi C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/espcoredump C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/esptool_py C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/fatfs C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/freertos C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/hal C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/heap C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/http_parser C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/idf_test C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/ieee802154 C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/json C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/log C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/lwip C:/Users/AndriiFilippov/runtime-launchConfiguration(12)/fcvijnl/main C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/mbedtls C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/mqtt C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/newlib C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/nvs_flash C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/openthread C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/partition_table C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/perfmon C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/protobuf-c C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/protocomm C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/pthread C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/sdmmc C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/soc C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/spi_flash C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/spiffs C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/tcp_transport C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/touch_element C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/ulp C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/unity C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/usb C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/vfs C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/wear_levelling C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/wifi_provisioning C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/wpa_supplicant C:/Users/AndriiFilippov/e/new/esp-idf-v5.0/components/xtensa
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/AndriiFilippov/runtime-launchConfiguration(12)/fcvijnl/build
cmake --build . -- -v
[1/855] cmd.exe /C "cd /D C:\Users\AndriiFilippov\runtime-launchConfiguration(12)\fcvijnl\build && C:\Users\AndriiFilippov\.espressif\tools\cmake\3.24.0\bin\cmake.exe -E touch C:/Users/AndriiFilippov/runtime-launchConfiguration(12)/fcvijnl/build/project_elf_src_esp32s2.c"
```

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/dialogs/BuildView.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+import org.eclipse.ui.part.ViewPart;
+
+import com.espressif.idf.core.build.ReHintPair;
+import com.espressif.idf.core.util.HintsUtil;
+
+public class BuildView extends ViewPart
+{
+
+	private TableViewer hintsTableViewer;
+	private final String[] titles = { "Error Message", "Hint" }; //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/dialogs/BuildView.java:37]: We can externalize these too!

**Review by kolipakakondal — APPROVED** [review]
LGTM

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/dialogs/BuildView.java (hunk 1 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/dialogs/BuildView.java]: done

**Comment by AndriiFilippov** [comment]
Hi @sigmaaa ! 

Tested under:
OS - Windows 10 

Issue still persist.

**Comment by kolipakakondal** [comment]
> Hi @sigmaaa !
> 
> Tested under: OS - Windows 10
> 
> Issue still persist.

Hi @sigmaaa Can you please check this issue? Let's target this for 2.11.0

**Comment by sigmaaa** [comment]
Hi @AndriiFilippov, @kolipakakondal,

fixed in the last commit by catching incompatible Python regex. This regex is fixed on a version higher than 5.0.3 on esp-idf side as well, that's why I was not able to reproduce this problem earlier

**Comment by sigmaaa** [comment]
Hi @kolipakakondal, @alirana01, @AndriiFilippov 
I also added the commit with the optimization of pattern compilation by refactoring the ReHintPair class. The idea is to compile all regular expressions only once. Please, take a look

**Comment by AndriiFilippov** [comment]
@sigmaaa hi ! 

Tested under:
OS - Windows 10 / MacOS / Linux Ubuntu 20.04

ESP-IDF: v5.0 - v5.1

release 5.1 / 5.1.1 / release 5.0 / 5.0.3 - able to enable / disable hint feature. able to see useful hints.

LGTM 👍

**Review by kolipakakondal — APPROVED** [review]
LGTM

**Pull request merged**